### PR TITLE
Feat/proposals db view

### PR DESCRIPTION
**Pull request by msadura**
### <samp>🤖 Generated by Copilot at 9955da0</samp>

This pull request enhances the storybook setup and the focalboard component for the app.charmverse.io repository. It updates the mock data and the mock API server to match the app logic and types, and adds drag and drop support and custom property types for focalboard tables. It also improves the code readability and consistency in some files.

### WHY
Proposals page db-like view

supported:
- single-view table layout
- sorting
- adding properties directly in table and editing them (only for admins)
- filtering by status and category
- column calculations
- hiding properties

Screenshot 2023-09-06 at 00 11 03

### CODE REVIEW

File: components/proposals/components/ProposalProperties/hooks/useProposalsBoardAdapter.ts [hunk 1]
```diff
@@ -1,63 +1,103 @@
...
   spaceId?: string;
 }) {
-  const proposalFields = proposal?.fields || { properties: {} };
+  const proposalFields = (proposal?.fields || { properties: {} }) as ProposalFields;
   const proposalSpaceId = proposal?.spaceId || spaceId || '';
 
-  return {
+  proposalFields.properties = {
+    ...proposalFields.properties,
+    // add default field values on the fly
```
Comment (msadura) [inline components/proposals/components/ProposalProperties/hooks/useProposalsBoardAdapter.ts:91]: mapping proposal "default" property values to focalboard properties constructed on the fly with specidif ids

File: components/proposals/components/ProposalsBoard/ProposalsBoard.tsx [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+    setUrlWithoutRerender(router.pathname, { id: pageId });
+    showProposal({
+      pageId,
+      onClose
+    });
+  }
+
+  return (
+    <Table
```
Comment (msadura) [inline components/proposals/components/ProposalsBoard/ProposalsBoard.tsx:31]: starting with just single  table view

File: components/proposals/components/ProposalsBoard/utils/boardData.ts [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+export function getDefaultBoard({
+  properties = [],
+  categories = [],
+  customOnly = false
+}: {
+  properties: IPropertyTemplate[] | undefined;
+  categories: ProposalCategory[] | undefined;
+  customOnly?: boolean;
+}) {
+  const board = createBoard({
```
Comment (msadura) [inline components/proposals/components/ProposalsBoard/utils/boardData.ts:28]: building mock board on the fly with default properties + custom ones

File: components/proposals/hooks/useProposals.tsx [hunk 4]
```diff
@@ -7,17 +10,49 @@ import { useCurrentSpace } from 'hooks/useCurrentSpace';
...
+  statusFilter: ProposalStatusFilter;
+  categoryIdFilter: string;
+  setStatusFilter: (status: ProposalStatusFilter) => void;
+  setCategoryIdFilter: (categoryId: string) => void;
+  mutateProposals: KeyedMutator<ProposalWithUsers[]>;
+  isLoading: boolean;
+  archiveProposal: (input: ArchiveProposalRequest) => Promise<void>;
+};
+
+export const ProposalsContext = createContext<Readonly<ProposalsContextType>>({
```
Comment (msadura) [inline components/proposals/hooks/useProposals.tsx:25]: Changed useProposals hook to context

File: components/common/BoardEditor/focalboard/src/mutator.ts [hunk 5]
```diff
@@ -36,6 +43,20 @@ export class Mutator {
 
   private undoDisplayId?: string;
 
+  private customMutatorUpdaters: MutatorUpdaters | null = null;
```
Comment (msadura) [inline components/common/BoardEditor/focalboard/src/mutator.ts:46]: a way to easily override hardcoded block patching  methods

File: components/common/BoardEditor/utils/blockUtils.ts [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import type { Block } from '@charmverse/core/prisma';
```
Comment (msadura) [inline components/common/BoardEditor/utils/blockUtils.ts:1]: utils just moved from charm client

File: components/[pageId]/DocumentPage/components/ProposalProperties.tsx [hunk 7]
```diff
@@ -108,6 +110,7 @@ export function ProposalProperties({
         refreshProposalPermissions()
       ]);
       mutateTasks();
+      mutateProposals();
```
Comment (mattcasey) [inline components/[pageId]/DocumentPage/components/ProposalProperties.tsx:113]: why does proposal properties require refreshing all proposals?

File: components/[pageId]/DocumentPage/components/ProposalProperties.tsx [hunk 8]
```diff
@@ -137,6 +140,7 @@ export function ProposalProperties({
     }
     refreshProposal();
     refreshProposalFlowFlags(); // needs to run when reviewers change?
+    mutateProposals();
```
Comment (mattcasey) [inline components/[pageId]/DocumentPage/components/ProposalProperties.tsx:143]: This is not great that we have to reload proposals on every update, but i know sockets would make this much more work.

In the future, I think we should follow other the pattern for page updates where we send a webhook event for proposal updates, and the context providers listen to the changes.

File: components/common/BoardEditor/focalboard/src/mutator.ts (hunk 5 above)
Comment (mattcasey) [reply components/common/BoardEditor/focalboard/src/mutator.ts:46]: clever.. and seems fine :)

File: components/common/stories/ProposalsPage/ProposalsPageStory.tsx [hunk 9]
```diff
@@ -0,0 +1,70 @@
...
+  createMockPage({
+    proposalId: p.id,
+    id: p.id,
+    type: 'proposal',
+    title: `A simple proposition ${i}`,
+    content: jsonDoc(_.p('This is the content'))
+  })
+);
+
+ProposalsPageStory.parameters = {
```
Comment (mattcasey) [inline components/common/stories/ProposalsPage/ProposalsPageStory.tsx:53]: Thanks for the new storybook page :)

**Review by mattcasey — CHANGES_REQUESTED** [review]
Seems to work really well! I loaded it up on to my local dev environment. For some reason the migration failed in staging, I just rolled it back and deleted your staging env for now. You could try running the migration against staging from your laptop to see why it failed?

Sort doesn't work for me. Not sure if it's the cause, but I see there's a network 500 error when I use it.  We are using __defaultView as a block id:
image

**Comment by msadura** [comment]
> Seems to work really well! I loaded it up on to my local dev environment. For some reason the migration failed in staging, I just rolled it back and deleted your staging env for now. You could try running the migration against staging from your laptop to see why it failed?
> 
> Sort doesn't work for me. Not sure if it's the cause, but I see there's a network 500 error when I use it. We are using __defaultView as a block id: image

ok found the reason for broken migration and pushing the fix.

As for issue with local env - seems like you did not run migration locally. I am usingids like __defaultView __defaultBoard for default stuff in proposals

**Pull request merged**